src/jit/decomposelongs.cpp

```diff
@@ -1056,24 +1055,22 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
                 }
                 else
                 {
-                    Range().Remove(loOp1);
-
                     assert(count >= 32);
 
-                    // Zero out loResult (shift of >= 32 bits shifts all lo bits to hiResult)
-                    loResult = m_compiler->gtNewZeroConNode(TYP_INT);
-                    Range().InsertBefore(tree, loResult);
-
                     if (count < 64)
                     {
                         if (count == 32)
                         {
                             // Move loOp1 into hiResult (shift of 32 bits is just a mov of lo to hi)
-                            hiResult = loOp1;
-                            Range().InsertBefore(tree, hiResult);
+                            LIR::Use loOp1Use(Range(), &gtLong->gtOp.gtOp1, gtLong);
+                            loOp1Use.ReplaceWithLclVar(m_compiler, m_blockWeight);
+
+                            hiResult = loOp1Use.Def();
+                            Range().Remove(gtLong);
                         }
                         else
                         {
+                            Range().Remove(gtLong);
```

Comment:
I see, we can compute hiResult first to avoid a tmp in this case.

---
In reply to: [86410153](https://github.com/dotnet/coreclr/pull/7972#discussion_r86410153) [](ancestors = 86410153)